### PR TITLE
fix #49235 Objects that pass the spread syntax can no longer be assigned to assignable types.

### LINKED ISSUE #49235: Objects that pass the spread syntax can no longer be assigned to assignable types.
# Bug Report

### 🔎 Search Terms

+ PropertyDescriptorMap
+ PropertyDescriptor
+ Spread

### 🕗 Version & Regression Information

- This has changed since version 4.6

### ⏯ Playground Link

Playground link with relevant code:

+ [Nightly (fail)](https://www.typescriptlang.org/play?ts=next#code/CYUwxgNghgTiAEYD2A7AzgF3gBxk7IMGAliGgIwBc8A3vANoAe1mMxKA5gLrUAKeBIgE8AImTBtsGJDHgBfANwAoJQHkARgCtwGAHSgAZuxD98hEmQAUNOQBocA86QoBKZUuTosdXb9xmiZwAmeXgAXgcAiwplDW0wPUNjU0Fo6ztI1OC3JSA)
+ [4.7.0-beta (fail)](https://www.typescriptlang.org/play?ts=4.7.0-beta#code/CYUwxgNghgTiAEYD2A7AzgF3gBxk7IMGAliGgIwBc8A3vANoAe1mMxKA5gLrUAKeBIgE8AImTBtsGJDHgBfANwAoJQHkARgCtwGAHSgAZuxD98hEmQAUNOQBocA86QoBKZUuTosdXb9xmiZwAmeXgAXgcAiwplDW0wPUNjU0Fo6ztI1OC3JSA)
+ [4.6.4 (fail)](https://www.typescriptlang.org/play?ts=4.6.4#code/CYUwxgNghgTiAEYD2A7AzgF3gBxk7IMGAliGgIwBc8A3vANoAe1mMxKA5gLrUAKeBIgE8AImTBtsGJDHgBfANwAoJQHkARgCtwGAHSgAZuxD98hEmQAUNOQBocA86QoBKZUuTosdXb9xmiZwAmeXgAXgcAiwplDW0wPUNjU0Fo6ztI1OC3JSA)
+ [4.5.5 (passed)](https://www.typescriptlang.org/play?ts=4.5.5#code/CYUwxgNghgTiAEYD2A7AzgF3gBxk7IMGAliGgIwBc8A3vANoAe1mMxKA5gLrUAKeBIgE8AImTBtsGJDHgBfANwAoJQHkARgCtwGAHSgAZuxD98hEmQAUNOQBocA86QoBKZUuTosdXb9xmiZwAmeXgAXgcAiwplDW0wPUNjU0Fo6ztI1OC3JSA)

### 💻 Code

```ts
declare const properties1: { [x: string]: PropertyDescriptor };

// This will pass. This is because the type '{ [x: string]: PropertyDescriptor }' can be assigned to the type 'PropertyDescriptorMap'.
Object.defineProperties({}, properties1);

// But when it passes the spread syntax, it fails even though the types are the same.
const { ...properties2 } = properties1;
Object.defineProperties({}, properties2);
```

### 🙁 Actual behavior

I get the following error:

```
Argument of type '{ [x: string]: PropertyDescriptor; }' is not assignable to parameter of type 'PropertyDescriptorMap & ThisType<any>'.
  Type '{ [x: string]: PropertyDescriptor; }' is not assignable to type 'PropertyDescriptorMap'.
    Index signature for type 'symbol' is missing in type '{ [x: string]: PropertyDescriptor; }'.
```

But, type `{ [x: string]: PropertyDescriptor; }` is assignable to type `PropertyDescriptorMap`. Objects copied with spread syntax cannot be assigned even though they have the same type.

### 🙂 Expected behavior

This type error does not occur. For the same type, it should be consistently assignable / non-assignable anywhere.

### PR DESCRIPTION
**Pull request by bentongxyz**
Fixes #49235

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -21497,13 +21497,17 @@ namespace ts {
...
+            const objectFlags = getObjectFlags(type);
             return type.flags & TypeFlags.Intersection
                 ? every((type as IntersectionType).types, isObjectTypeWithInferableIndex)
                 : !!(
                     type.symbol
                     && (type.symbol.flags & (SymbolFlags.ObjectLiteral | SymbolFlags.TypeLiteral | SymbolFlags.Enum | SymbolFlags.ValueModule)) !== 0
                     && !(type.symbol.flags & SymbolFlags.Class)
                     && !typeHasCallOrConstructSignatures(type)
+                ) || !!(
+                    objectFlags
```
Comment (sandersn) [inline src/compiler/checker.ts:21509]: getObjectFlags always returns a value, so you don't need to check for undefined (or 0)

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -21497,13 +21497,17 @@ namespace ts {
...
+                    && objectFlags & ObjectFlags.ObjectRestType
                 ) || !!(getObjectFlags(type) & ObjectFlags.ReverseMapped && isObjectTypeWithInferableIndex((type as ReverseMappedType).source));
```
Comment (sandersn) [inline src/compiler/checker.ts:21511]: Can you replace `getObjectFlags(type)` with `objectFlags` here too? getObjectFlags is likely to be inlined, so you could also call it from each place.

**Review by sandersn — CHANGES_REQUESTED** [review]
I think this is correct, although I'd like to get a second opinion from @gabritto. I have a couple of suggestions too.

**Comment by bentongxyz** [comment]
@sandersn I have made updates accordingly, thank you for the review :)

**Review by sandersn — APPROVED** [review]
I'll leave for a few days to see what @gabritto thinks.

**Review by gabritto — APPROVED** [review]

**Pull request merged**